chain/chunks/src/lib.rs

```diff
@@ -1202,79 +1226,163 @@ impl ShardsManager {
+    ///                          denotes whether the chunk header has been validated or not
+    /// `chain_head`: current chain head, used for validating chunk header
+    ///
+    /// Returns
+    ///  ProcessPartialEncodedChunkResult::Known: if all information in
+    ///    the `partial_encoded_chunk` is already known and no further processing is needed
+    ///  ProcessPartialEncodedChunkResult::NeedBlock: if the previous block is needed
+    ///    to finish processing
+    ///  ProcessPartialEncodedChunkResult::NeedMorePartsOrReceipts: if more parts and receipts
+    ///    are needed for processing the full chunk
+    ///  ProcessPartialEncodedChunkResult::HaveAllPartsAndReceipts: if all parts and
+    ///    receipts in the chunk are received and the chunk has been processed.
+    pub fn process_partial_encoded_chunk(
+        &mut self,
+        partial_encoded_chunk: MaybeValidated<&PartialEncodedChunkV2>,
+        chain_head: Option<&Tip>,
+        chain_store: &mut ChainStore,
+        rs: &mut ReedSolomonWrapper,
+    ) -> Result<ProcessPartialEncodedChunkResult, Error> {
         let header = &partial_encoded_chunk.header;
         let chunk_hash = header.chunk_hash();
-
-        // 2. Leave if we received known chunk
+        // Verify the partial encoded chunk is valid and worth processing
+        // 1.a Leave if we received known chunk
+        // Note that we have to still process the chunk even if all parts and receipts included
+        // in `partial_encoded_chunk` are already in `encoded_chunks`. This is because
+        // `try_process_full_chunk` can only succ
         if let Some(entry) = self.encoded_chunks.get(&chunk_hash) {
-            let know_all_parts = partial_encoded_chunk
-                .parts
-                .iter()
-                .all(|part_entry| entry.parts.contains_key(&part_entry.part_ord));
-
-            if know_all_parts {
-                let know_all_receipts = partial_encoded_chunk
-                    .receipts
-                    .iter()
-                    .all(|receipt| entry.receipts.contains_key(&receipt.1.to_shard_id));
-
-                if know_all_receipts {
-                    return Ok(ProcessPartialEncodedChunkResult::Known);
-                }
+            if entry.complete {
+                return Ok(ProcessPartialEncodedChunkResult::Known);
             }
         };
 
-        // 3. Checking chunk height
+        // 1.b Checking chunk height
         let chunk_requested = self.requested_partial_encoded_chunks.contains_key(&chunk_hash);
         if !chunk_requested {
             if !self.encoded_chunks.height_within_horizon(header.height_created()) {
-                return Err(Error::ChainError(ErrorKind::InvalidChunkHeight.into()));
+                return Ok(ProcessPartialEncodedChunkResult::Known);
             }
             // We shouldn't process unrequested chunk if we have seen one with same (height_created + shard_id)
```

Comment:
I didn't  fully undestand this case -- as this partial_encoded_chunk might contain different parts of the chunk that we didn't see yet, right ?

Reply:
This change is not relevant to this PR. I'll just remove it to avoid confusion. I just thought this error doesn't make sense and wanted to get rid of it. I'll do that in a different PR